swarm/src/behaviour/peer_addresses.rs

```diff
@@ -0,0 +1,329 @@
+use crate::behaviour::FromSwarm;
+use crate::{DialError, DialFailure, NewExternalAddrOfPeer};
+
+use libp2p_core::Multiaddr;
+use libp2p_identity::PeerId;
+
+use lru::LruCache;
+
+use std::collections::HashSet;
+use std::num::NonZeroUsize;
+
+/// Struct for tracking peers' external addresses of the [`Swarm`](crate::Swarm).
+#[derive(Debug)]
+pub struct PeerAddresses(LruCache<PeerId, HashSet<Multiaddr>>);
+
+impl PeerAddresses {
+    /// Creates cache with capacity of `cache_size`.
+    pub fn new(cache_size: NonZeroUsize) -> Self {
+        Self(LruCache::new(cache_size))
+    }
+
+    /// Feed a [`FromSwarm`] event to this struct.
+    ///
+    /// Returns whether the event changed peer's known external addresses.
+    pub fn on_swarm_event(&mut self, event: &FromSwarm) -> bool {
+        match event {
+            FromSwarm::NewExternalAddrOfPeer(NewExternalAddrOfPeer { peer_id, addr }) => {
+                self.add(*peer_id, (*addr).clone())
+            }
+            FromSwarm::DialFailure(DialFailure {
+                peer_id: Some(peer_id),
+                error: DialError::Transport(errors),
+                ..
+            }) => {
+                for (addr, _error) in errors {
+                    self.remove(peer_id, addr);
+                }
+                true
+            }
+            _ => false,
+        }
+    }
+
+    /// Adds address to cache.
+    /// Appends address to the existing set if peer addresses already exist.
+    /// Creates a new cache entry for peer_id if no addresses are present.
+    /// Returns true if the newly added address was not previously in the cache.
+    ///
+    pub fn add(&mut self, peer: PeerId, address: Multiaddr) -> bool {
+        let address = prepare_addr(&peer, &address);
+        if let Some(cached) = self.0.get_mut(&peer) {
+            cached.insert(address)
+        } else {
+            let mut set: HashSet<Multiaddr> = HashSet::new();
+            set.insert(address);
+            self.0.put(peer, set);
+            true
+        }
+    }
+
+    /// Returns peer's external addresses.
+    pub fn get(&mut self, peer: &PeerId) -> impl Iterator<Item = Multiaddr> {
+        self.0.get(peer).cloned().unwrap_or_default().into_iter()
+    }
+
+    /// Removes address from peer addresses cache.
+    /// Returns true if the address was removed.
+    pub fn remove(&mut self, peer: &PeerId, address: &Multiaddr) -> bool {
+        self.0.get_mut(peer).map_or_else(
+            || false,
+            |addrs| {
+                let address = prepare_addr(peer, address);
+                addrs.remove(&address)
+            },
+        )
```

Comment:
I find this confusing to read. I'd suggest we do a `match` instead and use an early return of `return false` if we don't have entry for the given peer.

Reply:
:heavy_check_mark: